riak_test/intercepts/intercept.erl

```diff
@@ -0,0 +1,392 @@
+%% -------------------------------------------------------------------
+%%
+%% Copyright (c) 2015 Basho Technologies, Inc.
+%%
+%% This file is provided to you under the Apache License,
+%% Version 2.0 (the "License"); you may not use this file
+%% except in compliance with the License.  You may obtain
+%% a copy of the License at
+%%
+%%   http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%%-------------------------------------------------------------------
+
+-module(intercept).
+%% Export explicit API but also send compile directive to export all
+%% because some of these private functions are useful in their own
+%% right.
+-export([add/3, add/4]).
+-compile(export_all).
+
+-type abstract_code() :: term().
+-type form() :: term().
+-type proplist(K, V) :: proplists:proplist(K, V).
+-type fun_name() :: atom().
+-type fun_type() :: fun_name() | tuple().
+-type target_fun() :: {fun_name(), arity()}.
+-type intercept_fun() :: fun_type().
+-type mapping() :: proplist(target_fun(), intercept_fun()).
+-type form_mod() :: fun((form()) -> form()).
+-type code_mod() :: fun((form(), abstract_code()) -> abstract_code()).
+
+%% The "original" is the `Target' module with the suffix `_orig'.  It
+%% is where original code for the `Target' module resides after
+%% intercepts are added.
+-define(ORIGINAL(Mod), list_to_atom(atom_to_list(Mod) ++ "_orig")).
+-define(FAKE_LINE_NO,1).
+
+%% @doc Add intercepts against the `Target' module.
+%%
+%% `Target' - The module on which to intercept calls.
+%%            E.g. `hashtree'.
+%%
+%% `Intercept' - The module containing intercept definitions.
+%%               E.g. `hashtree_intercepts'
+%%
+%% `Mapping' - The mapping from target functions to intercept
+%%             functions.
+%%
+%%             E.g. `[{{update_perform,2}, sleep_update_perform}]'
+-spec add(module(), module(), mapping()) -> ok.
+add(Target, Intercept, Mapping, OutDir) ->
```

Comment:
Agree, but I don't like to change intercepts.erl in riak_cs repo since it should remain just a copy of one in riak_test repo.